src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -199,44 +286,83 @@ public void PacketReset(SNIHandle handle, bool write, SNIPacket packet)
-            }
-
-            switch (serverNameParts[0])
+            else
             {
-                case TdsEnums.TCP:
-                    return CreateTcpHandle(serverNameParts[1], timerExpire, callbackObject, parallel);
-
-                case TdsEnums.NP:
-                    return CreateNpHandle(serverNameParts[1], timerExpire, callbackObject, parallel);
+                // when tcp protocol is specified
+                if (fullServerName.ToLower().StartsWith(TdsEnums.TCP + ":"))
+                {
+                    int protocolHeaderLength = TdsEnums.TCP.Length + 1; // including ':'
+                    string serverNameWithOutProtocol = fullServerName.Substring(protocolHeaderLength, fullServerName.Length - protocolHeaderLength);
+                    sniHandle = CreateTcpHandle(serverNameWithOutProtocol, timerExpire, callbackObject, parallel, ref spnBuffer, isIntegratedSecurity);
+                }
+                // when np protocol is specified
+                else if (fullServerName.ToLower().StartsWith(TdsEnums.NP + ":"))
+                {
+                    int protocolHeaderLength = TdsEnums.NP.Length + 1; // including ':'
+                    string serverNameWithOutProtocol = fullServerName.Substring(protocolHeaderLength, fullServerName.Length - protocolHeaderLength);
+                    sniHandle = CreateNpHandle(serverNameWithOutProtocol, timerExpire, callbackObject, parallel);
+                }
+                // possibly error case
+                else
+                {
+                    int portOrInstanceNameIndex = Math.Max(fullServerName.LastIndexOf(','), fullServerName.LastIndexOf('\\'));
+                    string serverNameWithOutPortOrInstanceName = portOrInstanceNameIndex > 0 ? fullServerName.Substring(0, portOrInstanceNameIndex) : fullServerName;
+                    IPAddress address = null;
 
-                default:
-                    if (parallel)
+                    // when no protocol is specified, and fullServerName is IPv6
+                    if (IPAddress.TryParse(serverNameWithOutPortOrInstanceName, out address) && address.AddressFamily == AddressFamily.InterNetworkV6)
                     {
-                        SNICommon.ReportSNIError(SNIProviders.INVALID_PROV, 0, SNICommon.MultiSubnetFailoverWithNonTcpProtocol, string.Empty);
+                        // default to using tcp if no protocol is provided
+                        sniHandle = CreateTcpHandle(fullServerName, timerExpire, callbackObject, parallel, ref spnBuffer, isIntegratedSecurity);
                     }
+                    // error case for sure
                     else
                     {
-                        SNICommon.ReportSNIError(SNIProviders.INVALID_PROV, 0, SNICommon.ProtocolNotSupportedError, string.Empty);
+                        // when invalid protocol is specified
+                        if (fullServerName.Split(':').Length == 2)
```

Comment:
It'd be better to avoid using Split for something like this, e.g.
```C#
private static bool OccursOnce(string s, char c)
{
    int pos = s.IndexOf(c);
    return pos >= 0 & s.IndexOf(c, pos+1) == -1;
}
...
if (OccursOnce(fullServerName, ':')) { ... }
```